test/handshake_helper.c

```diff
@@ -123,6 +124,48 @@ static int select_server_ctx(SSL *s, void *arg, int ignore)
     }
 }
 
+static int early_select_server_ctx(SSL *s, void *arg, int ignore)
+{
+    const char *servername;
+    PACKET pkt, list, name;
+    const unsigned char *p;
+    unsigned int type;
+    size_t len;
+    HANDSHAKE_EX_DATA *ex_data =
+        (HANDSHAKE_EX_DATA*)(SSL_get_ex_data(s, ex_data_idx));
+
+    if (!SSL_early_get0_ext(s, TLSEXT_TYPE_server_name, &p, &len) ||
+        !PACKET_buf_init(&pkt, p, len) ||
+        !PACKET_as_length_prefixed_2(&pkt, &list) ||
+        !PACKET_get_1(&list, &type) || type != TLSEXT_NAMETYPE_host_name ||
+        !PACKET_as_length_prefixed_2(&list, &name))
+        return 0;
+    servername = (const char *)PACKET_data(&name);
+    len = PACKET_remaining(&name);
+
+    if (len == strlen("server2") && strncmp(servername, "server2", len) == 0) {
+        SSL_CTX *new_ctx = arg;
+        SSL_set_SSL_CTX(s, new_ctx);
+        /*
+         * Copy over all the SSL_CTX options - reasonable behavior
+         * allows testing of cases where the options between two
+         * contexts differ/conflict
+         */
+        SSL_clear_options(s, 0xFFFFFFFFL);
+        SSL_set_options(s, SSL_CTX_get_options(new_ctx));
+
+        ex_data->servername = SSL_TEST_SERVERNAME_SERVER2;
+        return 1;
+    } else if (len == strlen("server1") &&
+               strncmp(servername, "server1", len) == 0) {
+        ex_data->servername = SSL_TEST_SERVERNAME_SERVER1;
+        return 1;
+    } else if (ignore) {
+        ex_data->servername = SSL_TEST_SERVERNAME_SERVER1;
+        return 1;
+    }
+    return 0;
+}
```

Comment:
Yeah, I think it's a bad idea for this...  it's better to test under conditions that our users will get into.